[MSE][GStreamer] Pause after seek is not working

calvaris opened the pull request:
#### f270bdb7b34c8245165f90be0c1dc8f864993ded
<pre>
[MSE][GStreamer] Pause after seek is not working
<a href="https://bugs.webkit.org/show_bug.cgi?id=263317">https://bugs.webkit.org/show_bug.cgi?id=263317</a>

Reviewed by Philippe Normand.

So far we are just asking the pipeline if we were paused or not but that does not work when the pipeline is
transitioning or seeking. That creates desynchronization between the media element and the player. We now consider the
pipeline in the final state while it is transitioning, as it can handle other requests while at it.

We also need to force ready state change when the pipeline finishes the state change to paused or playing because the
player will report state changes sooner.

This changes won&apos;t apply to MediaStream because the dynamics of prerolling are much different.

Fly by style change in isPipelineSeeking.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::isPipelineSeeking const):
(WebCore::MediaPlayerPrivateGStreamer::paused const):
(WebCore::MediaPlayerPrivateGStreamer::updateStates):

Canonical link: <a href="https://commits.webkit.org/274730@main">https://commits.webkit.org/274730@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f8ce70778e6ae47f24ce34542462523a7668a4b8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39850 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18861 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42225 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42394 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35752 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42157 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21765 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16191 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33215 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40424 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15897 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34458 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13755 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13781 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35414 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43673 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/36187 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35750 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39521 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14665 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12063 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37794 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16284 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8941 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16332 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/15928 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
```diff
@@ -470,19 +470,24 @@ bool MediaPlayerPrivateGStreamer::paused() const
         return false;
     }
 
+#if !defined(GST_DISABLE_GST_DEBUG) || !defined(NDEBUG) || defined(MEDIA_STREAM)
```
the mediastream check is incorrect (should be ENABLE_MEDIA_STREAM and check its value) but anyway... the check should runtime, using `isMediaStreamPlayer()` as mentioned earlier.

calvaris replied:
Do you mean I should remove the compile time checks and just keep the runtime one?

philn replied:
Move the runtime check before this ifdef is what i mean :)

calvaris replied:
Check last version, IMHO, this is the way to go!

calvaris committed:
[MSE][GStreamer] Pause after seek is not working
https://bugs.webkit.org/show_bug.cgi?id=263317

Reviewed by Philippe Normand.

So far we are just asking the pipeline if we were paused or not but that does not work when the pipeline is
transitioning or seeking. That creates desynchronization between the media element and the player. We now consider the
pipeline in the final state while it is transitioning, as it can handle other requests while at it.

We also need to force ready state change when the pipeline finishes the state change to paused or playing because the
player will report state changes sooner.

This changes won't apply to MediaStream because the dynamics of prerolling are much different.

Fly by style change in isPipelineSeeking.

* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::isPipelineSeeking const):
(WebCore::MediaPlayerPrivateGStreamer::paused const):
(WebCore::MediaPlayerPrivateGStreamer::updateStates):

Canonical link: https://commits.webkit.org/274730@main
